### ISSUE TITLE
Resubmit does not work

### ISSUE THREAD

**Issue by mbudiu-vmw** [p4lang/behavioral-model#556]
There are several reports that resubmit does not work. This may be a bmv2 issue, or a compiler issue. 
This is the compiler issue: https://github.com/p4lang/p4c/issues/1043 which also has some information about reproducing the problem.

**Comment by paulsubrata55** [comment]
I am also facing this issue with simple_switch_grpc: Added a code snippet for resubmit action:

    action ofpat_goto_table_(bit<7> table_id) {                                  
        meta.of_metadata.table_id = table_id;                                    
        resubmit(meta);                                                          
    }                                                                            
    table ofpat_goto_table {                                                     
        key = {                                                                  
            meta.of_metadata.index : exact;                                      
        }                                                                        
                                                                                 
        actions = {                                                              
            ofpat_goto_table_;                                                   
            NoAction;                                                            
        }                                                                        
        size = 1024;                                                             
    }

When there is a hit in this table getting the following logs from bmv2:

[12:59:05.191] [bmv2] [T] [thread 13478] [0.0] [cxt 0] Applying table 'ofpat_goto_table'
[12:59:05.191] [bmv2] [D] [thread 13478] [0.0] [cxt 0] Looking up key:
* of_metadata.index   : 00000001

[12:59:05.191] [bmv2] [D] [thread 13478] [0.0] [cxt 0] Table 'ofpat_goto_table': hit with handle 0
[12:59:05.191] [bmv2] [D] [thread 13478] [0.0] [cxt 0] Dumping entry 0
Match key:
* of_metadata.index   : EXACT     00000001
Action entry: ofpat_goto_table_ - 1,

[12:59:05.191] [bmv2] [D] [thread 13478] [0.0] [cxt 0] Action entry is ofpat_goto_table_ - 1,
[12:59:05.192] [bmv2] [T] [thread 13478] [0.0] [cxt 0] Action ofpat_goto_table_
[12:59:05.192] [bmv2] [T] [thread 13478] [0.0] [cxt 0] openflow_tables.p4(146) Primitive meta.of_metadata.table_id = table_id
[12:59:05.192] [bmv2] [T] [thread 13478] [0.0] [cxt 0] Primitive (no source info)
[12:59:05.192] [bmv2] [D] [thread 13478] [0.0] [cxt 0] Pipeline 'ingress': end
[12:59:05.192] [bmv2] [D] [thread 13478] [0.0] [cxt 0] Egress port is 0
[12:59:05.192] [bmv2] [D] [thread 13479] [0.0] [cxt 0] Pipeline 'egress': start
[12:59:05.192] [bmv2] [D] [thread 13479] [0.0] [cxt 0] Pipeline 'egress': end
[12:59:05.193] [bmv2] [D] [thread 13479] [0.0] [cxt 0] Deparser 'deparser': start
[12:59:05.193] [bmv2] [D] [thread 13479] [0.0] [cxt 0] Updating checksum 'cksum'
[12:59:05.193] [bmv2] [D] [thread 13479] [0.0] [cxt 0] Deparsing header 'ethernet'
[12:59:05.193] [bmv2] [D] [thread 13479] [0.0] [cxt 0] Deparsing header 'ipv4'
[12:59:05.193] [bmv2] [D] [thread 13479] [0.0] [cxt 0] Deparser 'deparser': end
[12:59:05.194] [bmv2] [D] [thread 13483] [0.0] [cxt 0] Transmitting packet of size 124 out of port 0

Seems like it did not find any primitive for the resubmit.

**Comment by antoninbas** [comment]
The name is misleading for resubmit_flag. It is actually not a flag, but is used to store the id of the field list to be resubmitted with the packet, with a value of 0 meaning "no resubmit". This is the JSON code generated by p4c for the resubmit primitive:
```
        {
          "op" : "resubmit",
          "parameters" : [
            {
              "type" : "hexstr",
              "value" : "0x0"
            }
          ]
        }
```
The primitive actually ends-up writing 0 to "resubmit_flag", which means no resubmit. So it has the same effect as not calling resubmit at all.
This is definitely an inconsistency in the bmv2 JSON format, as it requires field_lists ids to start at 1 not 0.

In other words, the "resubmit_flag" metadata field is actually used like the "lf_field_list" metadata field and the width should probably be > 1.

**Comment by antoninbas** [comment]
I'm "fixing" this in p4c.

**Comment by mbudiu-vmw** [comment]
If I understand right, the only thing that is needed is this:

```
    BMV2::nextId("field_lists");
```

in `Backend::convert`

**Comment by antoninbas** [comment]
Fixed with https://github.com/p4lang/p4c/pull/1183
